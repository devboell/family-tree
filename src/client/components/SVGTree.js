import React from 'react'
import { PropTypes } from 'prop-types'

import SVGNode from './SVGNode'


const boxWidth = 80
const marginX = 30
const boxHeight = 60
const marginY = 50
const nodeWidth = boxWidth + marginX
const nodeHeight = boxHeight + marginY


const SVGTree = ({ treeData }) => (
  <svg
    width={treeData.subTreeWidth + 10}
    height={((treeData.treeHeight + 1) * nodeHeight) + 10}
  >
    <SVGNode
      node={treeData}
      offsetX={0}
      linkTo={null}
      {...{ boxWidth, marginX, boxHeight, marginY }}
    />
  </svg>
)

SVGTree.propTypes = {
  treeData: PropTypes.object.isRequired,
}


export const transformWithPartners = (tree, level = -1) => {
  const node = {
    value: tree.name,
    level: level + 1,
    width: nodeWidth,
    treeHeight: 0,
    partners: [],
    subTreeWidth: 0,
  }
  tree.partners.forEach((partner) => {
    const partnerNode = {
      value: partner.partner.name,
      width: boxWidth,
      children: [],
      childrenWidth: [],
      subTreeWidth: 0,
    }
    // console.log('partner.children = ', partner.children)
    if (partner.children.length) {
      partner.children.forEach((child) => {
        // console.log('child = ', child)
        const childNode = transformWithPartners(child, node.level)
        // console.log('childNode = ', childNode)
        if (child.partners.length) {
          partnerNode.subTreeWidth += childNode.subTreeWidth
          partnerNode.childrenWidth.push(childNode.subTreeWidth)
        } else {
          partnerNode.subTreeWidth += childNode.width
          partnerNode.childrenWidth.push(childNode.width)
        }
        if (node.treeHeight <= childNode.treeHeight + 1) {
          node.treeHeight = childNode.treeHeight + 1
        }
        partnerNode.children.push(childNode)
      })
    }
    if (partnerNode.subTreeWidth < (nodeWidth * 2)) {
      partnerNode.subTreeWidth = (nodeWidth * 2)
    } // min width for partners
    node.partners.push(partnerNode)
    node.subTreeWidth += partnerNode.subTreeWidth
  })

  return node
}

/*
export const transformWithRelations = (tree, level = -1) => {
  const node = {
    value: tree.value,
    level: level + 1,
    width: nodeWidth,
    treeHeight: 0,
    relations: [],
    subTreeWidth: 0,
  }
  tree.relations.forEach((relation) => {
    const relationNode = {
      value: relation.value,
      width: boxWidth,
      children: [],
      childrenWidth: [],
      subTreeWidth: 0,
    }

    if (relation.children.length) {
      relation.children.forEach((child) => {
        const childNode = transformWithRelations(child, node.level)
        if (child.relations.length) {
          relationNode.subTreeWidth += childNode.subTreeWidth
          relationNode.childrenWidth.push(childNode.subTreeWidth)
        } else {
          relationNode.subTreeWidth += childNode.width
          relationNode.childrenWidth.push(childNode.width)
        }
        if (node.treeHeight <= childNode.treeHeight + 1) {
          node.treeHeight = childNode.treeHeight + 1
        }
        relationNode.children.push(childNode)
      })
    }
    if (relationNode.subTreeWidth < (nodeWidth * 2)) {
      relationNode.subTreeWidth = (nodeWidth * 2)
    } // min width for relations
    node.relations.push(relationNode)
    node.subTreeWidth += relationNode.subTreeWidth
  })

  return node
}
*/

export default SVGTree
