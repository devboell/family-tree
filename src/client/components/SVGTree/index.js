import React from 'react'
import { PropTypes } from 'prop-types'

import SVGNode from 'components/SVGNode'


const boxWidth = 80
const marginX = 30
const boxHeight = 60
const marginY = 50
const nodeWidth = boxWidth + marginX
const nodeHeight = boxHeight + marginY


const SVGTree = ({ treeData }) => (
  <svg
    width={treeData.subTreeWidth + 10}
    height={((treeData.treeHeight + 1) * nodeHeight) + 10}
  >
    <SVGNode
      node={treeData}
      offsetX={0}
      linkTo={null}
      {...{ boxWidth, marginX, boxHeight, marginY }}
    />
  </svg>
)

SVGTree.propTypes = {
  // TODO: https://github.com/facebook/react/issues/5676
  treeData: PropTypes.shape({}).isRequired,
}


export const transformWithPartners = (tree, level = -1) => {
  const node = {
    value: tree.name,
    level: level + 1,
    width: nodeWidth,
    treeHeight: 0,
    partners: [],
    subTreeWidth: 0,
  }
  tree.partners.forEach((partner) => {
    const partnerNode = {
      value: partner.partner.name,
      width: boxWidth,
      children: [],
      childrenWidth: [],
      subTreeWidth: 0,
    }
    if (partner.children.length) {
      partner.children.forEach((child) => {
        const childNode = transformWithPartners(child, node.level)
        if (child.partners.length) {
          partnerNode.subTreeWidth += childNode.subTreeWidth
          partnerNode.childrenWidth.push(childNode.subTreeWidth)
        } else {
          partnerNode.subTreeWidth += childNode.width
          partnerNode.childrenWidth.push(childNode.width)
        }
        if (node.treeHeight <= childNode.treeHeight + 1) {
          node.treeHeight = childNode.treeHeight + 1
        }
        partnerNode.children.push(childNode)
      })
    }
    if (partnerNode.subTreeWidth < (nodeWidth * 2)) {
      partnerNode.subTreeWidth = (nodeWidth * 2)
    } // min width for partners
    node.partners.push(partnerNode)
    node.subTreeWidth += partnerNode.subTreeWidth
  })

  return node
}


export default SVGTree
