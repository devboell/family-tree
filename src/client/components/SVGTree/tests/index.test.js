import React from 'react'
import SVGTree, { transformWithPartners } from '../index'


describe('SVGTree', () => {
  describe('transformWithPartners', () => {
    it('transforms root node tree', () => {
      const tree = {
        name: 'John',
        partners: [],
      }
      expect(transformWithPartners(tree)).toMatchSnapshot()
    })

    it('transforms root node with partner tree', () => {
      const tree = {
        name: 'John',
        partners: [
          {
            partner: {
              name: 'Mary',
            },
            children: [],
          },
        ],
      }
      expect(transformWithPartners(tree)).toMatchSnapshot()
    })

    it('transforms root node with partner and children', () => {
      const tree = {
        name: 'John',
        partners: [
          {
            partner: {
              name: 'Mary',
            },
            children: [
              {
                name: 'Tom',
                partners: [],
              },
              {
                name: 'Dick',
                partners: [],
              },
              {
                name: 'Harry',
                partners: [],
              },
            ],
          },
        ],
      }
      expect(transformWithPartners(tree)).toMatchSnapshot()
    })
  })
  describe('component', () => {
    it('renders a tree', () => {
      const tree = {
        name: 'John',
        partners: [
          {
            partner: {
              name: 'Mary',
            },
            children: [
              {
                name: 'Tom',
                partners: [],
              },
              {
                name: 'Dick',
                partners: [],
              },
              {
                name: 'Harry',
                partners: [],
              },
            ],
          },
        ],
      }
      const treeData = transformWithPartners(tree)
      const wrapper = mount(<SVGTree {...{ treeData }} />)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
