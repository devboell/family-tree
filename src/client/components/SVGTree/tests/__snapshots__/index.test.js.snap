// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGTree component renders a tree 1`] = `
<SVGTree
  treeData={
    Object {
      "level": 0,
      "partners": Array [
        Object {
          "children": Array [
            Object {
              "level": 1,
              "partners": Array [],
              "subTreeWidth": 0,
              "treeHeight": 0,
              "value": "Tom",
              "width": 110,
            },
            Object {
              "level": 1,
              "partners": Array [],
              "subTreeWidth": 0,
              "treeHeight": 0,
              "value": "Dick",
              "width": 110,
            },
            Object {
              "level": 1,
              "partners": Array [],
              "subTreeWidth": 0,
              "treeHeight": 0,
              "value": "Harry",
              "width": 110,
            },
          ],
          "childrenWidth": Array [
            110,
            110,
            110,
          ],
          "subTreeWidth": 330,
          "value": "Mary",
          "width": 80,
        },
      ],
      "subTreeWidth": 330,
      "treeHeight": 1,
      "value": "John",
      "width": 110,
    }
  }
>
  <svg
    height={230}
    width={340}
  >
    <SVGNode
      boxHeight={60}
      boxWidth={80}
      linkTo={null}
      marginX={30}
      marginY={50}
      node={
        Object {
          "level": 0,
          "partners": Array [
            Object {
              "children": Array [
                Object {
                  "level": 1,
                  "partners": Array [],
                  "subTreeWidth": 0,
                  "treeHeight": 0,
                  "value": "Tom",
                  "width": 110,
                },
                Object {
                  "level": 1,
                  "partners": Array [],
                  "subTreeWidth": 0,
                  "treeHeight": 0,
                  "value": "Dick",
                  "width": 110,
                },
                Object {
                  "level": 1,
                  "partners": Array [],
                  "subTreeWidth": 0,
                  "treeHeight": 0,
                  "value": "Harry",
                  "width": 110,
                },
              ],
              "childrenWidth": Array [
                110,
                110,
                110,
              ],
              "subTreeWidth": 330,
              "value": "Mary",
              "width": 80,
            },
          ],
          "subTreeWidth": 330,
          "treeHeight": 1,
          "value": "John",
          "width": 110,
        }
      }
      offsetX={0}
    >
      <g
        fill="none"
        stroke="black"
        strokeWidth="1"
        style={
          Object {
            "shapeRendering": "crispEdges",
          }
        }
      >
        <foreignObject
          height={60}
          width={80}
          x={85}
          y={0}
        >
          <styled.div
            height={60}
            width={80}
          >
            <div
              className="sc-bdVaJa iTJoaa"
              height={60}
              width={80}
            >
              <p>
                John
              </p>
            </div>
          </styled.div>
        </foreignObject>
        <g>
          <SVGRelationLink
            originX={195}
            originY={30}
            targetX={167}
          >
            <path
              d="M 195 30 H 167"
              stroke="red"
            />
          </SVGRelationLink>
          <SVGRelation
            boxHeight={60}
            boxWidth={80}
            marginX={30}
            marginY={50}
            offsetX={0}
            relationNode={
              Object {
                "children": Array [
                  Object {
                    "level": 1,
                    "partners": Array [],
                    "subTreeWidth": 0,
                    "treeHeight": 0,
                    "value": "Tom",
                    "width": 110,
                  },
                  Object {
                    "level": 1,
                    "partners": Array [],
                    "subTreeWidth": 0,
                    "treeHeight": 0,
                    "value": "Dick",
                    "width": 110,
                  },
                  Object {
                    "level": 1,
                    "partners": Array [],
                    "subTreeWidth": 0,
                    "treeHeight": 0,
                    "value": "Harry",
                    "width": 110,
                  },
                ],
                "childrenWidth": Array [
                  110,
                  110,
                  110,
                ],
                "subTreeWidth": 330,
                "value": "Mary",
                "width": 80,
              }
            }
            y={0}
          >
            <g>
              <foreignObject
                height={60}
                width={80}
                x={195}
                y={0}
              >
                <styled.div
                  height={60}
                  width={80}
                >
                  <div
                    className="sc-bdVaJa iTJoaa"
                    height={60}
                    width={80}
                  >
                    <p>
                      Mary
                    </p>
                  </div>
                </styled.div>
              </foreignObject>
              <SVGNode
                boxHeight={60}
                boxWidth={80}
                linkTo={
                  Object {
                    "x": 180,
                    "y": 30,
                  }
                }
                marginX={30}
                marginY={50}
                node={
                  Object {
                    "level": 1,
                    "partners": Array [],
                    "subTreeWidth": 0,
                    "treeHeight": 0,
                    "value": "Tom",
                    "width": 110,
                  }
                }
                offsetX={0}
              >
                <g
                  fill="none"
                  stroke="black"
                  strokeWidth="1"
                  style={
                    Object {
                      "shapeRendering": "crispEdges",
                    }
                  }
                >
                  <foreignObject
                    height={60}
                    width={80}
                    x={30}
                    y={110}
                  >
                    <styled.div
                      height={60}
                      width={80}
                    >
                      <div
                        className="sc-bdVaJa iTJoaa"
                        height={60}
                        width={80}
                      >
                        <p>
                          Tom
                        </p>
                      </div>
                    </styled.div>
                  </foreignObject>
                  <path
                    d="M 70 110 V 85 H 180 V 30"
                  />
                </g>
              </SVGNode>
              <SVGNode
                boxHeight={60}
                boxWidth={80}
                linkTo={
                  Object {
                    "x": 180,
                    "y": 30,
                  }
                }
                marginX={30}
                marginY={50}
                node={
                  Object {
                    "level": 1,
                    "partners": Array [],
                    "subTreeWidth": 0,
                    "treeHeight": 0,
                    "value": "Dick",
                    "width": 110,
                  }
                }
                offsetX={110}
              >
                <g
                  fill="none"
                  stroke="black"
                  strokeWidth="1"
                  style={
                    Object {
                      "shapeRendering": "crispEdges",
                    }
                  }
                >
                  <foreignObject
                    height={60}
                    width={80}
                    x={140}
                    y={110}
                  >
                    <styled.div
                      height={60}
                      width={80}
                    >
                      <div
                        className="sc-bdVaJa iTJoaa"
                        height={60}
                        width={80}
                      >
                        <p>
                          Dick
                        </p>
                      </div>
                    </styled.div>
                  </foreignObject>
                  <path
                    d="M 180 110 V 85 H 180 V 30"
                  />
                </g>
              </SVGNode>
              <SVGNode
                boxHeight={60}
                boxWidth={80}
                linkTo={
                  Object {
                    "x": 180,
                    "y": 30,
                  }
                }
                marginX={30}
                marginY={50}
                node={
                  Object {
                    "level": 1,
                    "partners": Array [],
                    "subTreeWidth": 0,
                    "treeHeight": 0,
                    "value": "Harry",
                    "width": 110,
                  }
                }
                offsetX={220}
              >
                <g
                  fill="none"
                  stroke="black"
                  strokeWidth="1"
                  style={
                    Object {
                      "shapeRendering": "crispEdges",
                    }
                  }
                >
                  <foreignObject
                    height={60}
                    width={80}
                    x={250}
                    y={110}
                  >
                    <styled.div
                      height={60}
                      width={80}
                    >
                      <div
                        className="sc-bdVaJa iTJoaa"
                        height={60}
                        width={80}
                      >
                        <p>
                          Harry
                        </p>
                      </div>
                    </styled.div>
                  </foreignObject>
                  <path
                    d="M 290 110 V 85 H 180 V 30"
                  />
                </g>
              </SVGNode>
            </g>
          </SVGRelation>
        </g>
      </g>
    </SVGNode>
  </svg>
</SVGTree>
`;

exports[`SVGTree transformWithPartners transforms root node tree 1`] = `
Object {
  "level": 0,
  "partners": Array [],
  "subTreeWidth": 0,
  "treeHeight": 0,
  "value": "John",
  "width": 110,
}
`;

exports[`SVGTree transformWithPartners transforms root node with partner and children 1`] = `
Object {
  "level": 0,
  "partners": Array [
    Object {
      "children": Array [
        Object {
          "level": 1,
          "partners": Array [],
          "subTreeWidth": 0,
          "treeHeight": 0,
          "value": "Tom",
          "width": 110,
        },
        Object {
          "level": 1,
          "partners": Array [],
          "subTreeWidth": 0,
          "treeHeight": 0,
          "value": "Dick",
          "width": 110,
        },
        Object {
          "level": 1,
          "partners": Array [],
          "subTreeWidth": 0,
          "treeHeight": 0,
          "value": "Harry",
          "width": 110,
        },
      ],
      "childrenWidth": Array [
        110,
        110,
        110,
      ],
      "subTreeWidth": 330,
      "value": "Mary",
      "width": 80,
    },
  ],
  "subTreeWidth": 330,
  "treeHeight": 1,
  "value": "John",
  "width": 110,
}
`;

exports[`SVGTree transformWithPartners transforms root node with partner tree 1`] = `
Object {
  "level": 0,
  "partners": Array [
    Object {
      "children": Array [],
      "childrenWidth": Array [],
      "subTreeWidth": 220,
      "value": "Mary",
      "width": 80,
    },
  ],
  "subTreeWidth": 220,
  "treeHeight": 0,
  "value": "John",
  "width": 110,
}
`;
